piper/keystroke.py

```diff
@@ -0,0 +1,128 @@
+        self._keycode = 0
+        self._keyval = 0
+        self._modifiers = []
+
+    @GObject.Property(type=str)
+    def accel(self):
+        """The accelerator of the current keystroke, as generated by
+        Gtk.accelerator_name(keyval, mask)."""
+        return self._accel
+
+    def set_key(self, keycode, keyval):
+        """Sets the given keycode and keyval as the key of this keystroke.
+
+        @param keycode The raw key code as reported by GDK, as int.
+        @param keyval The key, as int. See the `gdk/gdkkeysyms.h` header file
+                      for a complete list of GDK key codes.
+        """
+        # TODO: is a validity check necessary here? In principle, all uses of
+        # this method come from ButtonDialog.on_key_press, so they should all be
+        # valid. Otherwise, what does a validity check look like?
+        self._keycode = keycode
+        self._keyval = keyval
+        self._update_accel()
+
+    def add_modifier(self, keycode, keyval):
+        """Adds the given modifier keycode and keyval to the list of modifiers
+        of this keystroke.
+
+        @param keycode The raw key code, as int.
+        @param keyval The key, as int. See the `gdk/gdkkeysyms.h` header file
+                      for a complete list of GDK key codes.
+        """
+        # TODO: is a validity check necessary here? In principle, all uses of
+        # this method come from ButtonDialog.on_key_press, so they should all be
+        # valid. Otherwise, what does a validity check look like?
+        self._modifiers.append((keycode, keyval))
+        self._update_accel()
+
+    def clear(self):
+        """Clears the current keystroke."""
+        self._keycode = 0
+        self._keyval = 0
+        self._modifiers = []
+        self._update_accel()
+
+    def from_evdev(self, keycode, modifiers):
+        """Converts the given evdev keycodes to their GDK counterparts and sets
+        them as the current keystroke.
+
+        @param keycode The evdev keycode, as int.
+        @param modifiers A list of modifier keycodes, as [int].
+        """
+        # TODO: get keyval from keycode. Without this, the accelerator won't be
+        # printed correctly.
+        keycode += self._XORG_KEYCODE_OFFSET
+        keyval = 0
+        self.set_key(keycode, keyval)
+        for modifier in modifiers:
+            keycode = modifier + self._XORG_KEYCODE_OFFSET
+            keyval = 0
+            self.add_modifier(keycode, keyval)
+
+    def to_evdev(self):
+        """Returns a list of keycodes with the first being the pressed key and
+        the rest modifiers, if any. These keycodes map to those as defined by
+        evdev as opposed to those defined by GDK, thus the returned list is
+        directly usable by libratbag."""
+        if self._keycode > 0:
+            ret = [self._keycode - self._XORG_KEYCODE_OFFSET]
+        else:
+            # TODO: this gives KEY_RESERVED; what does libratbag do with that?
```

Comment:
This would happen because it's our default keycode value, so if the user attempts to set an initially emptry keystroke but then bails out with BackSpace, this method will be called on a keycode of value 0.